engine/execution/state/delta/view.go

```diff
@@ -48,6 +48,7 @@ func AlwaysEmptyGetRegisterFunc(owner, controller, key string) (flow.RegisterVal
 func NewView(readFunc GetRegisterFunc) *View {
 	return &View{
 		delta:             NewDelta(),
+		spockSecretLock:   &sync.Mutex{},
```

Comment:
What is it about this usage which makes the existing `sync.Mutex` field problematic? 

There are several other similar patterns in the code base, for example:
https://github.com/onflow/flow-go/blob/051bb593a2dc79e686b62a86dbf310d418d96375/engine/common/synchronization/request_heap.go#L14-L25

https://github.com/onflow/flow-go/blob/63d0eac7c47c7385adb008dde95c0639facea488/engine/consensus/approvals/chunk_collector.go#L11-L25

Reply:
This is only a problem if you copy value of the structure (or read it in a different goroutine after the write).  In the view case it was read by the `assert.Equal`.  Generally, having non-pointer `Mutex` slightly is faster (no dereference, hence no potential for a TLB walk) but there are cases where it triggers race detector.

If you want I can move the cases you've mentioned to use the pointer to be more error-proof.

Reply:
Can you provide the race detector command and the test case that triggers the race without this change, and won't trigger with this change?

Reply:
It fails down the line in the computation tests:
```CC=/usr/bin/clang CXX=/usr/bin/clang++ ~/tmp/go/bin/go test -tags=relic -race ./engine/execution/computation/computer```

with:
```
==================
WARNING: DATA RACE
Read at 0x00c000012800 by goroutine 26:
  reflect.typedmemmove()
      /home/rbtz/tmp/go/src/runtime/mbarrier.go:178 +0x0
  reflect.packEface()
      /home/rbtz/tmp/go/src/reflect/value.go:123 +0xae
  reflect.valueInterface()
      /home/rbtz/tmp/go/src/reflect/value.go:1394 +0x18a
  reflect.Value.Interface()
      /home/rbtz/tmp/go/src/reflect/value.go:1364 +0xc9
  fmt.(*pp).printValue()
      /home/rbtz/tmp/go/src/fmt/print.go:722 +0xca
  fmt.(*pp).printValue()
      /home/rbtz/tmp/go/src/fmt/print.go:876 +0x12be
  fmt.(*pp).printArg()
      /home/rbtz/tmp/go/src/fmt/print.go:712 +0xdf4
  fmt.(*pp).doPrintf()
      /home/rbtz/tmp/go/src/fmt/print.go:1122 +0xe38
  fmt.Sprintf()
      /home/rbtz/tmp/go/src/fmt/print.go:219 +0x67
  github.com/stretchr/testify/mock.Arguments.Diff()
      /home/rbtz/go/pkg/mod/github.com/stretchr/testify@v1.7.1-0.20210824115523-ab6dc3262822/mock/mock.go:803 +0x175
  github.com/stretchr/testify/mock.(*Mock).findExpectedCall()
      /home/rbtz/go/pkg/mod/github.com/stretchr/testify@v1.7.1-0.20210824115523-ab6dc3262822/mock/mock.go:295 +0x144
  github.com/stretchr/testify/mock.(*Mock).MethodCalled()
      /home/rbtz/go/pkg/mod/github.com/stretchr/testify@v1.7.1-0.20210824115523-ab6dc3262822/mock/mock.go:401 +0xb9
  github.com/stretchr/testify/mock.(*Mock).Called()
      /home/rbtz/go/pkg/mod/github.com/stretchr/testify@v1.7.1-0.20210824115523-ab6dc3262822/mock/mock.go:391 +0x1a4
  github.com/onflow/flow-go/engine/execution/computation/computer/mock.(*ViewCommitter).CommitView()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/mock/view_committer.go:21 +0x1a4
  github.com/onflow/flow-go/engine/execution/computation/computer.(*blockCommitter).Run()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer.go:434 +0x195
  github.com/onflow/flow-go/engine/execution/computation/computer.(*blockComputer).executeBlock.func3()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer.go:178 +0x39

Previous write at 0x00c000012800 by goroutine 24:
  sync/atomic.AddInt32()
      /home/rbtz/tmp/go/src/runtime/race_amd64.s:305 +0xb
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x1a
  github.com/onflow/flow-go/engine/execution/state/delta.(*View).SpockSecret()
      /home/rbtz/porn/flow-go/engine/execution/state/delta/view.go:255 +0xda
  github.com/onflow/flow-go/engine/execution/state/delta.(*View).MergeView()
      /home/rbtz/porn/flow-go/engine/execution/state/delta/view.go:218 +0x2bb
  github.com/onflow/flow-go/engine/execution/computation/computer.(*blockComputer).executeBlock()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer.go:197 +0x11d8
  github.com/onflow/flow-go/engine/execution/computation/computer.(*blockComputer).ExecuteBlock()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer.go:103 +0x48f
  github.com/onflow/flow-go/engine/execution/computation/computer_test.TestBlockExecutor_ExecuteBlock.func1()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer_test.go:84 +0x11ea
  testing.tRunner()
      /home/rbtz/tmp/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /home/rbtz/tmp/go/src/testing/testing.go:1306 +0x47

Goroutine 26 (running) created at:
  github.com/onflow/flow-go/engine/execution/computation/computer.(*blockComputer).executeBlock()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer.go:177 +0xe64
  github.com/onflow/flow-go/engine/execution/computation/computer.(*blockComputer).ExecuteBlock()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer.go:103 +0x48f
  github.com/onflow/flow-go/engine/execution/computation/computer_test.TestBlockExecutor_ExecuteBlock.func1()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer_test.go:84 +0x11ea
  testing.tRunner()
      /home/rbtz/tmp/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /home/rbtz/tmp/go/src/testing/testing.go:1306 +0x47

Goroutine 24 (running) created at:
  testing.(*T).Run()
      /home/rbtz/tmp/go/src/testing/testing.go:1306 +0x726
  github.com/onflow/flow-go/engine/execution/computation/computer_test.TestBlockExecutor_ExecuteBlock()
      /home/rbtz/porn/flow-go/engine/execution/computation/computer/computer_test.go:45 +0xcb
  testing.tRunner()
      /home/rbtz/tmp/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /home/rbtz/tmp/go/src/testing/testing.go:1306 +0x47
```